plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/JdkLdapClient.java

```diff
@@ -0,0 +1,248 @@
+
+    private final Map<String, String> basicEnvironment;
+    private final Optional<SSLContext> sslContext;
+
+    @Inject
+    public JdkLdapClient(LdapConfig ldapConfig)
+    {
+        String ldapUrl = requireNonNull(ldapConfig.getLdapUrl(), "ldapUrl is null");
+        if (ldapUrl.startsWith("ldap://")) {
+            log.warn("Passwords will be sent in the clear to the LDAP server. Please consider using SSL to connect.");
+        }
+
+        this.basicEnvironment = ImmutableMap.<String, String>builder()
+                .put(INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory")
+                .put(PROVIDER_URL, ldapUrl)
+                .put(REFERRAL, ldapConfig.isIgnoreReferrals() ? "ignore" : "follow")
+                .build();
+
+        this.sslContext = Optional.ofNullable(ldapConfig.getTrustCertificate())
+                .map(JdkLdapClient::createSslContext);
+    }
+
+    @Override
+    public void validatePassword(String userDistinguishedName, String password)
+            throws NamingException
+    {
+        createUserDirContext(userDistinguishedName, password).close();
+    }
+
+    @Override
+    public boolean isGroupMember(String searchBase, String groupSearch, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
+                CloseableSearchResults search = searchContext(searchBase, groupSearch, context)) {
+            return search.hasMore();
+        }
+    }
+
+    @Override
+    public Set<String> lookupUserDistinguishedNames(String searchBase, String searchFilter, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
+                CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
+            ImmutableSet.Builder<String> distinguishedNames = ImmutableSet.builder();
+            while (search.hasMore()) {
+                distinguishedNames.add(search.next().getNameInNamespace());
+            }
+            return distinguishedNames.build();
+        }
+    }
+
+    @Override
+    public Set<String> lookupUserGroups(String searchBase, String searchFilter, String contextUserDistinguishedName, String contextPassword)
+            throws NamingException
+    {
+        try (CloseableContext context = createUserDirContext(contextUserDistinguishedName, contextPassword);
+                CloseableSearchResults search = searchContext(searchBase, searchFilter, context)) {
+            ImmutableSet.Builder<String> groupNames = ImmutableSet.builder();
+            if (search.hasMore()) {
+                Attributes attributes = search.next().getAttributes();
+                Attribute memberOfAttribute = attributes.get("memberof");
```

Comment:
Can we make this configurable ?

Reply:
What if the LDAP Server doesn't have this `memberOf` overlay ??
Can we have a fallback mechanism where we could search for a set of groups and then search if the users exists in that group ?